pkg/kube/krt/join.go

```diff
@@ -272,73 +225,93 @@ func (j *join[T]) updateMergedCache(key, handlerID string, merged *T) mergedCach
 	} else {
 		updatedEntry = mergedCacheEntry[T]{current: merged}
 	}
-	j.mergedCache[mergedCacheKey{key: key, handlerID: handlerID}] = updatedEntry
+
+	cache[mergedCacheKey{key: key, handlerID: handlerID}] = updatedEntry
 	// It's probably simpler to just always set the collection entry multiple times
 	// TODO: Ensure old vlues don't stick around for too long and prevent garbage collection
-	j.mergedCache[mergedCacheKey{key: key}] = updatedEntry
+	cache[mergedCacheKey{key: key}] = updatedEntry
 	return updatedEntry
 }
 
+func (j *join[T]) updateMergedCache(key, handlerID string, merged *T) mergedCacheEntry[T] {
+	j.Lock()
+	defer j.Unlock()
+	return updateMergeCacheLocked(j.mergedCache, merged, key, handlerID)
+}
+
+func handleInnerCollectionEvent[T any](
+	handler func(o []Event[T]),
+	handlerID string,
+	getMergedForKey func(key string) *T,
+	updateMergedCache func(key, handlerID string, merged *T) mergedCacheEntry[T],
+) func(o []Event[T]) {
+	return func(events []Event[T]) {
+		mergedEvents := make([]Event[T], 0, len(events))
+		changedKeys := make(map[string]struct{}, len(events))
+		// When we calculate the merged value for a given key, we're looking at the present state
+		// of our set of collections, not the state at the time of the event. Therefore, it's possible
+		// that the event state is stale and the merged value is different. Therefore, we should just
+		// operate on the key and not the event itself.
+		for _, i := range events {
+			key := GetKey(i.Latest())
+			if _, ok := changedKeys[key]; ok {
+				continue
+			}
+			changedKeys[key] = struct{}{}
+		}
+		// Loop through all of the keys that changed and create an event based on the currente state
+		// and our cached entries.
+		for key := range changedKeys {
+			merged := getMergedForKey(key)
+			entry := updateMergedCache(key, handlerID, merged)
+			var e Event[T]
+
+			switch {
+			case entry.current == nil && entry.prev == nil:
+				msg := "Merged (nested) join collection: Received event for key %s in handler %s but it's not longer in our set of collections. Skipping..."
+				log.Warnf(msg, key, handlerID)
```

Comment:
Should this be a warn or a debug? I think it's pretty cryptic for end-users to see. No strong feelings here.